include/gsl/gsl-lite.hpp

```diff
@@ -1540,32 +1560,54 @@ class not_null
     }
 #endif // gsl_HAVE( RVALUE_REFERENCE )
 
-#if gsl_HAVE( FUNCTION_REF_QUALIFIER )
-# if gsl_CONFIG( NOT_NULL_TRANSPARENT_GET )
-    gsl_api gsl_constexpr14 element_type &  get() const &  { return checked_ptr().get(); }
-    gsl_api gsl_constexpr14 element_type && get() &&       { return std::move(checked_ptr()).get(); }
-# else
-    gsl_api gsl_constexpr14 get_result_t get() const { return checked_ptr(); }
-# endif
-
-    gsl_api gsl_constexpr14 operator T() const & { return checked_ptr(); }
-    gsl_api gsl_constexpr14 operator T() &&      { return std::move(checked_ptr()); }
+#if gsl_CONFIG( NOT_NULL_TRANSPARENT_GET )
+    gsl_api gsl_constexpr14 element_type* get() const { return checked_ptr().get(); }
+#else
+    gsl_api gsl_constexpr14 get_result_t  get() const { return checked_ptr(); }
+#endif
 
-    gsl_api gsl_constexpr14 T const & operator->() const & { return checked_ptr(); }
-    gsl_api gsl_constexpr14 T &&      operator->() &&      { return std::move(checked_ptr()); }
+    // Ideally, we would convert to either `T const &` or `T &&` so we would permit a move but still avoid potentially unnecessary copies.
+    // Unfortunately, that would render code like `std::unique_ptr<T> p = not_null<std::unique_ptr<T>>(...);` ambiguous.
+    // We choose to always convert to `T` instead if function ref qualifiers are available, expecting that the cost of the copy will usually be
+    // alleviated by move semantics.
+    // This implies we cannot convert to const refs of move-only types; but that scenario should be rare.
 
-    gsl_api gsl_constexpr14 element_type &  operator*() const & { return *checked_ptr(); }
-    gsl_api gsl_constexpr14 element_type && operator*() &&      { return *std::move(checked_ptr()); }
+#if gsl_HAVE( RVALUE_REFERENCE ) && gsl_HAVE( TYPE_TRAITS ) && gsl_HAVE( DEFAULT_FUNCTION_TEMPLATE_ARG ) && gsl_HAVE( EXPLICIT )
+    // explicit conversion operator
+    // if type_traits is not available, then we can't distinguish is_convertible and is_constructible, so we use this unconditionally
 
-#else
-    gsl_api gsl_constexpr14 get_result_t get() const { return checked_ptr(); }
+    template< class U
+        gsl_REQUIRES_A(( std::is_constructible<U, T>::value && !std::is_convertible<T, U>::value && !gsl::detail::is_not_null_oracle<U>::value ))
+    >
+    gsl_api gsl_constexpr14 explicit operator U gsl_not_null_CONVERSION_REF() const gsl_not_null_LVALUE_REF { return checked_ptr(); }
+# if gsl_HAVE( FUNCTION_REF_QUALIFIER )
+    template< class U
+        gsl_REQUIRES_A(( std::is_constructible<U, T>::value && !std::is_convertible<T, U>::value && !gsl::detail::is_not_null_oracle<U>::value ))
+    >
+    gsl_api gsl_constexpr14 explicit operator U() && { return std::move(checked_ptr()); }
+# endif
 
-    gsl_api gsl_constexpr   operator get_result_t  () const   { return checked_ptr(); }
+    // implicit conversion operator if type_traits is available
```

Comment:
This comment is at least true, but it was only used originally as a contrast with the previous comment. I don't think it has a function here.

Reply:
Well, it *does* summarize the meaning of the 9 following unreadable lines quite well.